compiler/rustc_parse/src/parser/path.rs

```diff
@@ -373,7 +374,52 @@ impl<'a> Parser<'a> {
                     .into()
                 } else {
                     // `(T, U) -> R`
-                    let (inputs, _) = self.parse_paren_comma_seq(|p| p.parse_ty())?;
+                    let prev_token_before_parsing = self.prev_token.clone();
+
+                    let mut snapshot = None;
+                    if self.may_recover()
+                        && (self.token.can_begin_expr() || self.token.can_begin_pattern())
+                    {
+                        snapshot = Some(self.create_snapshot_for_diagnostic());
+                    }
+
+                    let (inputs, _) = match self.parse_paren_comma_seq(|p| p.parse_ty()) {
+                        Ok(output) => output,
+                        Err(mut error) if prev_token_before_parsing.kind == token::ModSep => {
+                            error.span_label(prev_token_before_parsing.span.with_hi(prev_token_before_parsing.span.hi() + BytePos(1)), "while parsing this list of parenthesized type arguments starting here");
+
+                            if let Some(mut snapshot) = snapshot {
+                                if ((style == PathStyle::Expr
+                                    && snapshot.parse_paren_comma_seq(|p| p.parse_expr()).is_ok())
+                                    || (style == PathStyle::Pat
+                                        && snapshot
+                                            .parse_paren_comma_seq(|p| {
+                                                p.parse_pat_allow_top_alt(
+                                                    None,
+                                                    RecoverComma::No,
+                                                    RecoverColon::No,
+                                                    CommaRecoveryMode::LikelyTuple,
+                                                )
+                                            })
+                                            .is_ok()))
+                                    && snapshot.token.kind != token::ModSep
+                                    && snapshot.token.kind != token::RArrow
+                                {
+                                    {
+                                        error.span_suggestion_verbose(
+                                            prev_token_before_parsing.span,
+                                            "consider removing `::`",
```

Comment:
The suggestion is already pretty good but I'd like to more context provided with the message. Something like `` consider removing `::` to make this a function call``. *Function call* is definitely not the right term here, maybe *call expression* / *tuple struct pattern* depending on the `style`.

Hopefully you can find a better more general term here, I don't know how to best call “$expr($expr, …)” / `$path($path, …)` lol.